### ISSUE TITLE
TF stopped working for me, until restarting browsers, and other thoughts

### ISSUE THREAD

**Issue by mick-p1982** [tabfloater/tabfloater.github.io#28]
This is such an obvious app, I've wondered for years why it's not built into Windows and today found your app because I was curious why Firefox limited its "picture-in-picture" mode to video.

Just some feedback and anecdote: First TF somehow stopped working for me (keeping its window on top) but of course I could quit out of the browsers and start them over again to get it back to working order. It would help I think if the companion app was more visible (systray? Start menu?) so we can tell if it's crashed or something. Is it a service or regular app?

More issues: 1) The minimize button seems to lose the floated windows. They're not in the taskbar but they do a minimize style animation. I assume again they belong to the "companion" app that's not represented in the taskbar. I think you can parent them to the browser even if they don't belong to it, but I'm not 100% certain. 2) I feel like the anchor tab definitely should not be removed. If it's not desirable to copy its contents then it should somehow become a placeholder tab. Anyway, I would prefer a copy given no option. Finally, in theory pressing Alt+I to anchor the floating tab should select/focus the anchor tab in the main browser window, but it might be less disorienting if the tab was never changed in the first place. If so it might make sense to not change tabs when anchoring.

More thoughts: I feel a single hotkey to toggle would be better. And it would be better if multiple tabs could be floated and the button state in the browser would be unique to every tab. Then the single hotkey would be analogous to pressing that button. If minimizing can't minimize to the taskbar then it should anchor that window instead.

My main use case for this is I think having more than one monitor is pretty silly with widescreen monitors, so I've always wanted a way to pull up a document I'm using for reference over the app I'm working in as if it was a tool window in the app I'm interacting with so I don't have to clumsily resize the app or shove it half off screen, etc. I don't really see the need for the corner positioning hotkeys in that case. It seems over complex to me.

**Comment by ba32107** [comment]
Hi, I actually only noticed this issue by accident. This repo is for the tabfloater.io website, the proper place for error reports is https://github.com/tabfloater/tabfloater. Please open any future issues there. 

Thanks for your feedback. Let me reply to each of your points.

> First TF somehow stopped working for me (keeping its window on top) but of course I could quit out of the browsers and start them over again to get it back to working order

I'm not able to figure out what the problem is just from this description alone. Please enable logging in the options, and have a look at the logs - maybe it will give you a better idea of what's wrong. I'm happy to help with this, if you need it.

> It would help I think if the companion app was more visible (systray? Start menu?) so we can tell if it's crashed or something. Is it a service or regular app?

The app is a console application that is started and stopped automatically by the browser each time you click the float button. It is not running continuously, therefore it never "crashes" by itself. If the browser fails to start the app, it gives you an error message that you can check in the Extensions section (depending on your browser). A systray icon wouldn't make sense because the app literally just runs for a few milliseconds at a time.

> The minimize button seems to lose the floated windows. They're not in the taskbar but they do a minimize style animation. I assume again they belong to the "companion" app that's not represented in the taskbar. I think you can parent them to the browser even if they don't belong to it, but I'm not 100% certain.

This is by design. The floating tab is supposed to be a "panel" within the main browser window. So, when you minimize the main window, the floating tab will minimize with it. Ideally, you're not even supposed to be able to minimize just the floating window, but there's no way to get rid of the title bar. If you accidentally minimize the floating window, you can restore it by minimizing your main browser window and restoring it - the floating tab will also restore. The floating tab actually does belong to the browser window, not the companion app.

> I feel like the anchor tab definitely should not be removed. If it's not desirable to copy its contents then it should somehow become a placeholder tab. Anyway, I would prefer a copy given no option. Finally, in theory pressing Alt+I to anchor the floating tab should select/focus the anchor tab in the main browser window, but it might be less disorienting if the tab was never changed in the first place. If so it might make sense to not change tabs when anchoring.

Sorry, I don't understand this at all. What is an "anchor tab"? I'm very confused by this whole paragraph - can you rephrase please?

> I feel a single hotkey to toggle would be better.

Fair point, although the idea is that with Fixed Positioning, you can move around the window with the four keys, just as if they were the arrows. The idea is that you "bring down" a tab, then "move it up", with the down/up buttons.

> And it would be better if multiple tabs could be floated and the button state in the browser would be unique to every tab.

Supporting multiple floating tabs is something I thought about, but it's a lot of work and I don't have a lot of time to work on this. Maybe for 2.0, if that ever happens.

> My main use case for this is I think having more than one monitor is pretty silly with widescreen monitors, so I've always wanted a way to pull up a document I'm using for reference over the app I'm working in as if it was a tool window in the app I'm interacting with so I don't have to clumsily resize the app or shove it half off screen, etc. I don't really see the need for the corner positioning hotkeys in that case. It seems over complex to me.

Again I don't quite understand you here, sorry. I want to highlight two features that might help you achieve a more optimal workflow:
 1. You can set the floating tab to be always on top, even on top of all other applications, not just the browsers
 2. You can set your own custom position and size for the floating tab, you don't need to use the built-in positioning methods. This is a brand new feature I just released half an hour ago. You should be receiving an update soon.



As I've answered your questions and remarks, I'm gonna close this issue now. If you need any additional help, or have any new ideas, please head over to the main repository (link above) and open a separate issue for each problem / feature request. Having a lot of different ideas in one issue like this makes it difficult to reply to them.

**Comment by mick-p1982** [comment]
> This is by design. The floating tab is supposed to be a "panel" within the main browser window. So, when you minimize the main window, the floating tab will minimize with it. Ideally, you're not even supposed to be able to minimize just the floating window, but there's no way to get rid of the title bar. If you accidentally minimize the floating window, you can restore it by minimizing your main browser window and restoring it - the floating tab will also restore. The floating tab actually does belong to the browser window, not the companion app.

Thanks for responding. For the record, on Windows, the only way to recover the tab after minimizing is to click the icon button for it in the browser. My impression was the window was being lost (inaccessible via the taskbar) in which case it needs a window that is visible on the taskbar. All windows need an icon on the taskbar...

The trouble I have (my main use case is to use it with other apps--honestly I can't imagine it being useful with a browser, but I suppose if you use browser based apps perhaps) is I find myself having to shuffle the top-most window around to keep it out of my way, since I can't minimize to the taskbar like a real window, and it causes anxiety to remember that if I do hit that button it will be lost (I know now how to recover it, by the button, but all-in-all this isn't a good experience, I mean I think it has a long way to grow.)

BTW, the stuff about the "anchor tab" means the original tab that spawned it. I do find the idea of popping it out of the browser is too bewildering. Just like when you use Firefox's picture-in-picture system, it doesn't remove the original webpage, and so when you close the PIP window you still have the original page/tab there, and it's not disorienting. In your case (TF) the original window disappears, and a random tab appears in its place... and upon closing TF it just becomes another seemingly random background tab, not a visible window. If it must pop-out I suggest using the same size as the tab's window for the pop-out. By being small it just means it has to be manually resized to get back tot the the original view. 

P.S. My apologies for landing on the wrong project page, no need to reply, I'm happy to provide input. In Win32 making a window appear in the taskbar and hiding the minimize/maximize buttons is trivial (since you mention this) by adjusting its window's style/ex mask. If your companion app creates the window it can do that, although a functioning minimize button would be a great improvement, so I recommend only adding to it WS_EX_APPWINDOW status.

**Comment by ba32107** [comment]
I find it difficult to understand what your use-case is with TabFloater. You said that you don't want the tab to pop out from the browser, and you also want the tab to be the same size as the original browser window. This goes against the whole point of TabFloater: it's supposed to mimic picture-and-picture (which by definition means a smaller window inside a larger one). So maybe what you're really after is some sort of [tab layout manager instead](https://chrome.google.com/webstore/detail/tab-resize-split-screen-l/bkpenclhmiealbebdopglffmfdiilejc)? Perhaps I misunderstood you - but I still don't quite see what you're trying to achieve.

> In your case (TF) the original window disappears, and a random tab appears in its place

This is the only way of implementing what I wanted. The browser API doesn't let you create "panels", so I had to work around it and come up with the best approximation I could.

> My impression was the window was being lost (inaccessible via the taskbar)

This is a fair point. If you had this problem, others might have it too. Hiding the floating window from the taskbar is actually done on purpose, to me, it made sense at the time. But it might be better to not do it, as it would prevent some confusion. I'll have a think about this.

> In Win32 making a window appear in the taskbar and hiding the minimize/maximize buttons is trivial

Yep, I'm aware of how to do it. The problem is, on Windows, Chrome actually doesn't use a "normal" window. Instead, Chrome actually draws the minimize/maximize/close buttons itself, and doesn't use the built-in, standard buttons that Windows provides. Therefore, if you try to hide those buttons via the Win32 API, nothing happens because they weren't there in the first place. At least I wasn't able to do it - if you can, please share a code snippet.

**Comment by mick-p1982** [comment]
> I find it difficult to understand what your use-case is with TabFloater. You said that you don't want the tab to pop out from the browser, and you also want the tab to be the same size as the original browser window. This goes against the whole point of TabFloater: it's supposed to mimic picture-and-picture (which by definition means a smaller window inside a larger one). So maybe what you're really after is some sort of tab layout manager instead? Perhaps I misunderstood you - but I still don't quite see what you're trying to achieve.

I think maybe you have a blind spot from working on TF too much. My use case is to see a webpage over an app that isn't the browser itself. Therefore it would have to "pop out" just as PnP does. But PnP doesn't remove the tab from the browser. It just removes the framed video. You just need to think of the tab/viewport in the browser as the frame as in the PnP example. (That means either open a new page in the floated tab so that there's two copies, or leave a blank viewport `<HTML />` in the browser so it's not disorienting. You're thinking about in terms of rearranging the window like in an IDE. That's the wrong way to think about it. Think about it like PnP.

> This is a fair point. If you had this problem, others might have it too. Hiding the floating window from the taskbar is actually done on purpose, to me, it made sense at the time. But it might be better to not do it, as it would prevent some confusion. I'll have a think about this.

It seems like a programming mistake unless the goal is only get the window back via the TF button in the browser toolbar. I can confirm that's the case for Firefox. But I would say also don't close the window with the X (Close) button, but return it to the browser instead of minimizing to the taskbar, I.e. unfloat it. Closing a PnP window doesn't close the webpage.

> Yep, I'm aware of how to do it. The problem is, on Windows, Chrome actually doesn't use a "normal" window. Instead, Chrome actually draws the minimize/maximize/close buttons itself, and doesn't use the built-in, standard buttons that Windows provides. Therefore, if you try to hide those buttons via the Win32 API, nothing happens because they weren't there in the first place. At least I wasn't able to do it - if you can, please share a code snippet.

I would make your own window (I assumed that's what the companion app does) and just make it a native system window (don't worry about mimicking Chrome) and hide the frame on the browser child window (assuming that's possible) and parent it to the new window you made. If you can't parent it, then just float it in front of it. Of course all this is assuming Windows 10 doesn't actively prevent you from managing cross process windows this way and Firefox and Chrome are content to let you do this kind of thing. I think it is common practice to reparent browser windows (there used to be a lot of "Grease Monkey" add-ons that would do this, if I'm no mistaken.)

**Comment by ba32107** [comment]
I do understand your use case now. While it's not what TabFloater is intended for, you are most welcome to use it this way. As mentioned earlier, there is now a custom positioning feature that might help you avoid moving the window around.

Thanks for your other suggestions. However, I'm not planning to do any major refactor or change in functionality.
